### PR TITLE
Kill globals and generalize Backend traits

### PR DESCRIPTION
**Pull request by clux**
This is meant to complete #7 once it's done.

### CODE REVIEW

**Comment by clux** [comment]
Got a few breaking changes now:

- `lal query` lists versions in reverse direction now (so people using `| tail` will be screwed)
- manifest.json now requires an extra `supportedEnvironments` key in each repo
- `lal query` without an `-e env` flag will crash (TODO: want to handle this better)
- `lal export` without an `-e env` flag will crash (TODO: want to handle this better)

But beyond that, things work as before, just with some more logging in `update`

File: src/export.rs [hunk 1]
```diff
@@ -9,14 +9,14 @@ pub fn export<T: CachedBackend + ?Sized>(
...
-    env: Option<&str>,
+    env: &str,
 ) -> LalResult<()> {
     if comp.to_lowercase() != comp {
         return Err(CliError::InvalidComponentName(comp.into()))
     }
 
     let dir = output.unwrap_or(".");
-    info!("Export {} {} to {}", env.unwrap_or("global"), comp, dir);
+    info!("Export {} {} to {}", env, comp, dir);
```
Comment (bencord0) [inline src/export.rs:27]: Yeah, this is much nicer now.

File: src/main.rs [hunk 2]
```diff
@@ -39,14 +39,14 @@ fn handle_manifest_agnostic_cmds(
         lal::export(backend,
                     a.value_of("component").unwrap(),
                     a.value_of("output"),
-                    explicit_env)
+                    explicit_env.unwrap()) // TODO: enforce unwrap condition
```
Comment (bencord0) [inline src/main.rs]: Do you want to handle these `unwrap()` now? or clean them up in another PR?

File: src/storage/download.rs [hunk 3]
```diff
@@ -94,21 +93,41 @@ impl<T: ?Sized> CachedBackend for T
 where
     T: Backend,
 {
+    /// Get the latest version of a component across all supported environments
```
Comment (bencord0) [inline src/storage/download.rs]: Add the note from #7 about how this returns a version which is available in all (supported) environments.
Even though a higher version is available, if it isn't everywhere, it won't be chosen.

File: src/storage/traits.rs [hunk 4]
```diff
@@ -38,25 +40,23 @@ pub struct Component {
...
-        name: &str,
-        version: Option<u32>,
-        loc: Option<&str>,
-    ) -> LalResult<Component>;
+    /// Get the version and location information of a component
+    ///
+    /// If no version is given, figure out what latest is
+    fn get_component_info(&self, name: &str, ver: Option<u32>, loc: &str) -> LalResult<Component>;
 
     /// Publish a release build (ARTIFACT dir) to a specific location
```
Comment (bencord0) [inline src/storage/traits.rs]: `/// Publish a release build's ARTIFACT to a backend specific location.`
If we're renaming the method, maybe just `publish_artifact`, with a note that this means everything in the directory, the tarball and the lockfile (and maybe a signature when we eventually support those).

File: src/update.rs [hunk 5]
```diff
@@ -50,7 +50,19 @@ pub fn update<T: CachedBackend + ?Sized>(
...
+
+            // First, since this potentially goes in the manifest
+            // make sure the version is found for all supported environments:
+            let res = backend.get_latest_supported_versions(comp, manifest.supportedEnvironments.clone())?;
+            let ver = match res.into_iter().max() {
+                Some(m) => m,
+                None => {
+                    return Err(CliError::NoIntersectedVersion(comp.clone()));
+                }
+            };
```
Comment (bencord0) [inline src/update.rs]: Sometimes I just enjoy reading rust error handlers.
It's something about knowing that after this line, `ver` is something sensible.

**Review by bencord0 — APPROVED** [review]

File: src/update.rs (hunk 5 above)
Comment (clux) [reply src/update.rs]: If you liked that, replace that entire match with a chained on:

```sh
.ok_or(CliError::NoIntersectedVersion(comp.clone()))?
```

Comment (bencord0) [reply src/update.rs]: ... tbh, I prefer it the other way.

**Review by bencord0 — APPROVED** [review]

File: src/update.rs (hunk 5 above)
Comment (clux) [reply src/update.rs]: hah, aw

File: src/main.rs [hunk 6]
```diff
@@ -39,10 +39,10 @@ fn handle_manifest_agnostic_cmds(
         lal::export(backend,
                     a.value_of("component").unwrap(),
                     a.value_of("output"),
-                    explicit_env.unwrap()) // TODO: enforce unwrap condition
+                    explicit_env)
```
Comment (bencord0) [inline src/main.rs]: cool

File: SPEC.md [hunk 7]
```diff
@@ -27,6 +27,7 @@ A per-repo file. Format looks like this (here annotated with illegal comments):
 {
   "name": "libwebsockets",  // name of repo
   "environment": "centos",  // name of environment found in the lal config
+  "supportedEnvironments": ["centos", "xenial"],
```
Comment (bencord0) [inline SPEC.md:30]: w00t

**Review by bencord0 — APPROVED** [review]

**Pull request merged**